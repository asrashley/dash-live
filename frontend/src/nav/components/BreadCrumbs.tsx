import { useCallback } from "preact/hooks";
import { useBreadcrumbs, type BreadCrumbItem } from "../hooks/useBreadcrumbs";

interface CrumbItemProps extends BreadCrumbItem {
  setLocation: (url: string) => void;
}

function CrumbItem({ title, href, active, id, setLocation }: CrumbItemProps) {
  const followInternalLink = useCallback(
    (ev: Event) => {
      const elt = ev.target as HTMLAnchorElement;
      const href = elt.getAttribute("href");
      ev.preventDefault();
      setLocation(href);
      return false;
    },
    [setLocation]
  );

  const body = active ? (
    <span>{title}</span>
  ) : (
    <a href={href} title={title} onClick={followInternalLink}>
      {title}
    </a>
  );
  const className = `breadcrumb-item ${active ? "active" : ""}`;
  return (
    <li id={id} className={className}>
      {body}
    </li>
  );
}

export function BreadCrumbs() {
  const { breadcrumbs, location, setLocation } = useBreadcrumbs();

  return (
    <nav
      className="breadcrumbs"
      aria-label="breadcrumb"
      data-location={location}
    >
      <ol className="breadcrumb">
        {breadcrumbs.map((crumb) => (
          <CrumbItem key={crumb.id} setLocation={setLocation} {...crumb} />
        ))}
      </ol>
    </nav>
  );
}
