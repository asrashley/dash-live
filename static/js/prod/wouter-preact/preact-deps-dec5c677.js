import { useState, useLayoutEffect, useEffect, useRef } from 'preact/hooks';
import 'preact';

// Copied from:
// https://github.com/facebook/react/blob/main/packages/shared/ExecutionEnvironment.js
const canUseDOM = !!(
  typeof window !== "undefined" &&
  typeof window.document !== "undefined" &&
  typeof window.document.createElement !== "undefined"
);

// TODO: switch to `export { useSyncExternalStore } from "preact/compat"` once we update Preact to >= 10.11.3
function is(x, y) {
  return (x === y && (x !== 0 || 1 / x === 1 / y)) || (x !== x && y !== y);
}
function useSyncExternalStore(subscribe, getSnapshot, getSSRSnapshot) {
  if (getSSRSnapshot && !canUseDOM) getSnapshot = getSSRSnapshot;
  const value = getSnapshot();

  const [{ _instance }, forceUpdate] = useState({
    _instance: { _value: value, _getSnapshot: getSnapshot },
  });

  useLayoutEffect(() => {
    _instance._value = value;
    _instance._getSnapshot = getSnapshot;

    if (!is(_instance._value, getSnapshot())) {
      forceUpdate({ _instance });
    }
  }, [subscribe, value, getSnapshot]);

  useEffect(() => {
    if (!is(_instance._value, _instance._getSnapshot())) {
      forceUpdate({ _instance });
    }

    return subscribe(() => {
      if (!is(_instance._value, _instance._getSnapshot())) {
        forceUpdate({ _instance });
      }
    });
  }, [subscribe]);

  return value;
}

// provide forwardRef stub for preact
function forwardRef(component) {
  return component;
}

// Userland polyfill while we wait for the forthcoming
// https://github.com/reactjs/rfcs/blob/useevent/text/0000-useevent.md
// Note: "A high-fidelty polyfill for useEvent is not possible because
// there is no lifecycle or Hook in React that we can use to switch
// .current at the right timing."
// So we will have to make do with this "close enough" approach for now.
const useEvent = (fn) => {
  const ref = useRef([fn, (...args) => ref[0](...args)]).current;
  useLayoutEffect(() => {
    ref[0] = fn;
  });
  return ref[1];
};

export { useSyncExternalStore as a, forwardRef as f, useEvent as u };
