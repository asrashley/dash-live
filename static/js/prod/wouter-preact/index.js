import { parse } from 'regexparam';
import { useBrowserLocation, useSearch as useSearch$1 } from './use-browser-location.js';
import { f as forwardRef, u as useEvent } from './preact-deps-dec5c677.js';
import { createContext, createElement, isValidElement, cloneElement, Fragment } from 'preact';
import { useContext, useRef, useLayoutEffect } from 'preact/hooks';

/*
 * Transforms `path` into its relative `base` version
 * If base isn't part of the path provided returns absolute path e.g. `~/app`
 */
const _relativePath = (base, path) =>
  !path.toLowerCase().indexOf(base.toLowerCase())
    ? path.slice(base.length) || "/"
    : "~" + path;

/**
 * When basepath is `undefined` or '/' it is ignored (we assume it's empty string)
 */
const baseDefaults = (base = "") => (base === "/" ? "" : base);

const absolutePath = (to, base) =>
  to[0] === "~" ? to.slice(1) : baseDefaults(base) + to;

const relativePath = (base = "", path) =>
  _relativePath(unescape(baseDefaults(base)), unescape(path));

/*
 * Removes leading question mark
 */
const stripQm = (str) => (str[0] === "?" ? str.slice(1) : str);

/*
 * decodes escape sequences such as %20
 */
const unescape = (str) => {
  try {
    return decodeURI(str);
  } catch (_e) {
    // fail-safe mode: if string can't be decoded do nothing
    return str;
  }
};

const sanitizeSearch = (search) => unescape(stripQm(search));

/*
 * Router and router context. Router is a lightweight object that represents the current
 * routing options: how location is managed, base path etc.
 *
 * There is a default router present for most of the use cases, however it can be overridden
 * via the <Router /> component.
 */

const defaultRouter = {
  hook: useBrowserLocation,
  searchHook: useSearch$1,
  parser: parse,
  base: "",
  // this option is used to override the current location during SSR
  ssrPath: undefined,
  ssrSearch: undefined,
  // customizes how `href` props are transformed for <Link />
  hrefs: (x) => x,
};

const RouterCtx = createContext(defaultRouter);

// gets the closest parent router from the context
const useRouter = () => useContext(RouterCtx);

/**
 * Parameters context. Used by `useParams()` to get the
 * matched params from the innermost `Route` component.
 */

const Params0 = {},
  ParamsCtx = createContext(Params0);

const useParams = () => useContext(ParamsCtx);

/*
 * Part 1, Hooks API: useRoute and useLocation
 */

// Internal version of useLocation to avoid redundant useRouter calls

const useLocationFromRouter = (router) => {
  const [location, navigate] = router.hook(router);

  // the function reference should stay the same between re-renders, so that
  // it can be passed down as an element prop without any performance concerns.
  // (This is achieved via `useEvent`.)
  return [
    relativePath(router.base, location),
    useEvent((to, navOpts) => navigate(absolutePath(to, router.base), navOpts)),
  ];
};

const useLocation = () => useLocationFromRouter(useRouter());

const useSearch = () => {
  const router = useRouter();
  return sanitizeSearch(router.searchHook(router));
};

const matchRoute = (parser, route, path, loose) => {
  // if the input is a regexp, skip parsing
  const { pattern, keys } =
    route instanceof RegExp
      ? { keys: false, pattern: route }
      : parser(route || "*", loose);

  // array destructuring loses keys, so this is done in two steps
  const result = pattern.exec(path) || [];

  // when parser is in "loose" mode, `$base` is equal to the
  // first part of the route that matches the pattern
  // (e.g. for pattern `/a/:b` and path `/a/1/2/3` the `$base` is `a/1`)
  // we use this for route nesting
  const [$base, ...matches] = result;

  return $base !== undefined
    ? [
        true,

        (() => {
          // for regex paths, `keys` will always be false

          // an object with parameters matched, e.g. { foo: "bar" } for "/:foo"
          // we "zip" two arrays here to construct the object
          // ["foo"], ["bar"] â†’ { foo: "bar" }
          const groups =
            keys !== false
              ? Object.fromEntries(keys.map((key, i) => [key, matches[i]]))
              : result.groups;

          // convert the array to an instance of object
          // this makes it easier to integrate with the existing param implementation
          let obj = { ...matches };

          // merge named capture groups with matches array
          groups && Object.assign(obj, groups);

          return obj;
        })(),

        // the third value if only present when parser is in "loose" mode,
        // so that we can extract the base path for nested routes
        ...(loose ? [$base] : []),
      ]
    : [false, null];
};

const useRoute = (pattern) =>
  matchRoute(useRouter().parser, pattern, useLocation()[0]);

/*
 * Part 2, Low Carb Router API: Router, Route, Link, Switch
 */

const Router = ({ children, ...props }) => {
  // the router we will inherit from - it is the closest router in the tree,
  // unless the custom `hook` is provided (in that case it's the default one)
  const parent_ = useRouter();
  const parent = props.hook ? defaultRouter : parent_;

  // holds to the context value: the router object
  let value = parent;

  // when `ssrPath` contains a `?` character, we can extract the search from it
  const [path, search] = props.ssrPath?.split("?") ?? [];
  if (search) (props.ssrSearch = search), (props.ssrPath = path);

  // hooks can define their own `href` formatter (e.g. for hash location)
  props.hrefs = props.hrefs ?? props.hook?.hrefs;

  // what is happening below: to avoid unnecessary rerenders in child components,
  // we ensure that the router object reference is stable, unless there are any
  // changes that require reload (e.g. `base` prop changes -> all components that
  // get the router from the context should rerender, even if the component is memoized).
  // the expected behaviour is:
  //
  //   1) when the resulted router is no different from the parent, use parent
  //   2) if the custom `hook` prop is provided, we always inherit from the
  //      default router instead. this resets all previously overridden options.
  //   3) when the router is customized here, it should stay stable between renders
  let ref = useRef({}),
    prev = ref.current,
    next = prev;

  for (let k in parent) {
    const option =
      k === "base"
        ? /* base is special case, it is appended to the parent's base */
          parent[k] + (props[k] || "")
        : props[k] || parent[k];

    if (prev === next && option !== next[k]) {
      ref.current = next = { ...next };
    }

    next[k] = option;

    // the new router is no different from the parent, use parent
    if (option !== parent[k]) value = next;
  }

  return createElement(RouterCtx.Provider, { value, children });
};

const h_route = ({ children, component }, params) => {
  // React-Router style `component` prop
  if (component) return createElement(component, { params });

  // support render prop or plain children
  return typeof children === "function" ? children(params) : children;
};

// a hook to cache the params object between renders (if they are shallow equal)
const useCachedParams = (value) => {
  let prev = useRef(Params0),
    curr = prev.current;

  for (const k in value) if (value[k] !== curr[k]) curr = value;
  if (Object.keys(value).length === 0) curr = value;
  return (prev.current = curr);
};

const Route = ({ path, nest, match, ...renderProps }) => {
  const router = useRouter();
  const [location] = useLocationFromRouter(router);

  const [matches, routeParams, base] =
    // `match` is a special prop to give up control to the parent,
    // it is used by the `Switch` to avoid double matching
    match ?? matchRoute(router.parser, path, location, nest);

  // when `routeParams` is `null` (there was no match), the argument
  // below becomes {...null} = {}, see the Object Spread specs
  // https://tc39.es/proposal-object-rest-spread/#AbstractOperations-CopyDataProperties
  const params = useCachedParams({ ...useParams(), ...routeParams });

  if (!matches) return null;

  const children = base
    ? createElement(Router, { base }, h_route(renderProps, params))
    : h_route(renderProps, params);

  return createElement(ParamsCtx.Provider, { value: params, children });
};

const Link = forwardRef((props, ref) => {
  const router = useRouter();
  const [currentPath, navigate] = useLocationFromRouter(router);

  const {
    to = "",
    href: targetPath = to,
    onClick: _onClick,
    asChild,
    children,
    className: cls,
    /* eslint-disable no-unused-vars */
    replace /* ignore nav props */,
    state /* ignore nav props */,
    /* eslint-enable no-unused-vars */

    ...restProps
  } = props;

  const onClick = useEvent((event) => {
    // ignores the navigation when clicked using right mouse button or
    // by holding a special modifier key: ctrl, command, win, alt, shift
    if (
      event.ctrlKey ||
      event.metaKey ||
      event.altKey ||
      event.shiftKey ||
      event.button !== 0
    )
      return;

    _onClick?.(event);
    if (!event.defaultPrevented) {
      event.preventDefault();
      navigate(targetPath, props);
    }
  });

  // handle nested routers and absolute paths
  const href = router.hrefs(
    targetPath[0] === "~" ? targetPath.slice(1) : router.base + targetPath,
    router // pass router as a second argument for convinience
  );

  return asChild && isValidElement(children)
    ? cloneElement(children, { onClick, href })
    : createElement("a", {
        ...restProps,
        onClick,
        href,
        // `className` can be a function to apply the class if this link is active
        className: cls?.call ? cls(currentPath === targetPath) : cls,
        children,
        ref,
      });
});

const flattenChildren = (children) =>
  Array.isArray(children)
    ? children.flatMap((c) =>
        flattenChildren(c && c.type === Fragment ? c.props.children : c)
      )
    : [children];

const Switch = ({ children, location }) => {
  const router = useRouter();
  const [originalLocation] = useLocationFromRouter(router);

  for (const element of flattenChildren(children)) {
    let match = 0;

    if (
      isValidElement(element) &&
      // we don't require an element to be of type Route,
      // but we do require it to contain a truthy `path` prop.
      // this allows to use different components that wrap Route
      // inside of a switch, for example <AnimatedRoute />.
      (match = matchRoute(
        router.parser,
        element.props.path,
        location || originalLocation,
        element.props.nest
      ))[0]
    )
      return cloneElement(element, { match });
  }

  return null;
};

const Redirect = (props) => {
  const { to, href = to } = props;
  const [, navigate] = useLocation();
  const redirect = useEvent(() => navigate(to || href, props));

  // redirect is guaranteed to be stable since it is returned from useEvent
  useLayoutEffect(() => {
    redirect();
  }, []); // eslint-disable-line react-hooks/exhaustive-deps

  return null;
};

export { Link, Redirect, Route, Router, Switch, matchRoute, useLocation, useParams, useRoute, useRouter, useSearch };
