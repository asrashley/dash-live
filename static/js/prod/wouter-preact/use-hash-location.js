import { a as useSyncExternalStore } from './preact-deps-dec5c677.js';
import 'preact/hooks';
import 'preact';

// array of callback subscribed to hash updates
const listeners = {
  v: [],
};

const onHashChange = () => listeners.v.forEach((cb) => cb());

// we subscribe to `hashchange` only once when needed to guarantee that
// all listeners are called synchronously
const subscribeToHashUpdates = (callback) => {
  if (listeners.v.push(callback) === 1)
    addEventListener("hashchange", onHashChange);

  return () => {
    listeners.v = listeners.v.filter((i) => i !== callback);
    if (!listeners.v.length) removeEventListener("hashchange", onHashChange);
  };
};

// leading '#' is ignored, leading '/' is optional
const currentHashLocation = () => "/" + location.hash.replace(/^#?\/?/, "");

const navigate = (to, { state = null } = {}) => {
  // calling `replaceState` allows us to set the history
  // state without creating an extra entry
  const [hash, search] = to.replace(/^#?\/?/, "").split("?");

  history.replaceState(
    state,
    "",
    // keep the current pathname, but replace query string and hash
    location.pathname +
      (search ? `?${search}` : location.search) +
      // update location hash, this will cause `hashchange` event to fire
      // normalise the value before updating, so it's always preceeded with "#/"
      (location.hash = `#/${hash}`)
  );
};

const useHashLocation = ({ ssrPath = "/" } = {}) => [
  useSyncExternalStore(
    subscribeToHashUpdates,
    currentHashLocation,
    () => ssrPath
  ),
  navigate,
];

useHashLocation.hrefs = (href) => "#" + href;

export { navigate, useHashLocation };
