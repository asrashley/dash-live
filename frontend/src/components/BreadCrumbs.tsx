import { useEffect } from 'preact/hooks'
import { navigate } from "wouter-preact/use-browser-location";

import { useBreadcrumbs } from "../hooks/useBreadcrumbs";

function followInternalLink(ev) {
  const href = ev.target.getAttribute('href');
  ev.preventDefault();
  navigate(href, {replace: true});
  return false;
}

/*
NOTE: this returns an HTML string, *not* markup objects, as it is used to
directly manipulate a DOM element outside of the Preact app.
*/
function breadCrumbItemHtml({title, href, active, id}) {
  let body;

  if (active) {
    body = title;
  } else {
    body = `<a href="${ href }" title="${title}">${ title }</a>`;
  }
  return `<li id="${id}" class="breadcrumb-item ${active ? 'active': ''}">${body}</li>`;
}
export function BreadCrumbs() {
  const { breadcrumbs, location } = useBreadcrumbs();
  const payload = breadcrumbs.map(crumb => breadCrumbItemHtml(crumb)).join('');

  const destElt = document.querySelector('header > .breadcrumbs > ol.breadcrumb');
  if (destElt) {
    destElt.innerHTML = payload;
  }

  useEffect(() => {
    for (const crumb of breadcrumbs) {
        document.getElementById(crumb.id)?.addEventListener('click', followInternalLink);
    }
    return () => {
      const anchors = document.querySelectorAll('header > .breadcrumbs .breadcrumb-item > a');
      for (let i=0; i < anchors.length; ++i){
        anchors.item(i).removeEventListener('click', followInternalLink);
      }
    };
  })

  return <div style="display:none" data-location={location} />;
}
