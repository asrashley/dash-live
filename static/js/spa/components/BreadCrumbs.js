import { html } from "htm/preact";
import { useEffect } from 'preact/hooks'
import { useLocation, useRouter } from 'wouter-preact';
import { navigate } from "wouter-preact/use-browser-location";

function followInternalLink(ev) {
  const href = ev.target.getAttribute('href');
  ev.preventDefault();
  navigate(href, {replace: true});
  return false;
}

/*
NOTE: this returns an HTML string, *not* markup objects, as it is used to
directly manipulate a DOM element outside of the Preact app.
*/
function breadCrumbItemHtml({title, href, active, id}) {
  let body;

  if (active) {
    body = title;
  } else {
    body = `<a href="${ href }" title="${title}">${ title }</a>`;
  }
  return `<li id="${id}" class="breadcrumb-item ${active ? 'active': ''}">${body}</li>`;
}

export const useBreadcrumbs = () => {
  const [location, setLocation] = useLocation();
  const { base } = useRouter();

  const baseParts = base === "" ? [""] : base.split('/');
  const parts = location === "/" ? [""] : location.split('/').slice(baseParts.length - 1);

  const breadcrumbs = parts.map((title, idx) => {
    const href = idx === 0 ? `${base}/` : `${base}${parts.slice(0, idx + 1).join('/')}`;
    const active = idx == (parts.length - 1);
    const useLink = idx > 0;
    const id = `crumb_${idx}`;
    if (title === '') {
      title = 'Home';
    }
    return {title, active, href, useLink, id};
  });

  return {breadcrumbs, location, setLocation};
};

export function BreadCrumbs() {
  const { breadcrumbs, location } = useBreadcrumbs();
  const payload = breadcrumbs.map(crumb => breadCrumbItemHtml(crumb)).join('');

  document.querySelector('header > .breadcrumbs > ol.breadcrumb').innerHTML = payload;

  useEffect(() => {
    for (const crumb of breadcrumbs) {
      if (crumb.useLink) {
        document.getElementById(crumb.id)?.addEventListener('click', followInternalLink);
      }
    }
    return () => {
      for (const elt of document.querySelectorAll('header > .breadcrumbs .breadcrumb-item > a')) {
        elt.removeEventListener('click', followInternalLink);
      }
    };
  })

  return html`<div style="display:none" data-location="${location}" />`;
}
