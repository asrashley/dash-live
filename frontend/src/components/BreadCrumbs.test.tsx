import { describe, expect, test } from "vitest";
import { Route } from "wouter-preact";

import { renderWithProviders } from "../test/renderWithProviders";
import { elementAsFragment } from '../test/asFragment';
import { BreadCrumbs } from "./BreadCrumbs";

describe("BreadCrumbs", () => {
  test("should display Breadcrumbs", () => {
    document.body.innerHTML = `<header>
          <div class="breadcrumbs">
            <ol class="breadcrumb" />
          </div>
        </header><div id="app" />`;
    const baseElement = document.getElementById("app");
    expect(baseElement).toBeDefined();
    expect(baseElement).not.toBeNull();
    renderWithProviders(
      <div>
        <BreadCrumbs />
        <Route path="/multi-period-streams/:stream">
          <div />
        </Route>
      </div>,
      {
        baseElement,
        path: "/multi-period-streams/.add",
      }
    );
    expect(document.querySelector("ol.breadcrumb")?.innerHTML).toBe(
      '<li id="crumb_0" class="breadcrumb-item "><a href="/" title="Home">Home</a></li>' +
        '<li id="crumb_1" class="breadcrumb-item "><a href="/multi-period-streams" ' +
        'title="multi-period-streams">multi-period-streams</a></li>' +
        '<li id="crumb_2" class="breadcrumb-item active">.add</li>'
    );
    expect(elementAsFragment(document.body)).toMatchSnapshot();
  });
});
