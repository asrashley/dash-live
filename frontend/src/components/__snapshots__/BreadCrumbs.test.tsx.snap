// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BreadCrumbs > should display Breadcrumbs for / 1`] = `
<DocumentFragment>
  <div>
    <nav
      aria-label="breadcrumb"
      class="breadcrumbs"
      data-location="/"
    >
      <ol
        class="breadcrumb"
      >
        <li
          class="breadcrumb-item active"
          id="crumb_0"
        >
          <span>
            Home
          </span>
        </li>
      </ol>
    </nav>
    <div>
      home page
    </div>
  </div>
</DocumentFragment>
`;

exports[`BreadCrumbs > should display Breadcrumbs for /multi-period-streams 1`] = `
<DocumentFragment>
  <div>
    <nav
      aria-label="breadcrumb"
      class="breadcrumbs"
      data-location="/multi-period-streams"
    >
      <ol
        class="breadcrumb"
      >
        <li
          class="breadcrumb-item "
          id="crumb_0"
        >
          <a
            href="/"
            title="Home"
          >
            Home
          </a>
        </li>
        <li
          class="breadcrumb-item active"
          id="crumb_1"
        >
          <span>
            multi-period-streams
          </span>
        </li>
      </ol>
    </nav>
    <div>
      list MPS
    </div>
  </div>
</DocumentFragment>
`;

exports[`BreadCrumbs > should display Breadcrumbs for /multi-period-streams/.add 1`] = `
<DocumentFragment>
  <div>
    <nav
      aria-label="breadcrumb"
      class="breadcrumbs"
      data-location="/multi-period-streams/.add"
    >
      <ol
        class="breadcrumb"
      >
        <li
          class="breadcrumb-item "
          id="crumb_0"
        >
          <a
            href="/"
            title="Home"
          >
            Home
          </a>
        </li>
        <li
          class="breadcrumb-item "
          id="crumb_1"
        >
          <a
            href="/multi-period-streams"
            title="multi-period-streams"
          >
            multi-period-streams
          </a>
        </li>
        <li
          class="breadcrumb-item active"
          id="crumb_2"
        >
          <span>
            .add
          </span>
        </li>
      </ol>
    </nav>
    <div>
      edit MPS
    </div>
  </div>
</DocumentFragment>
`;
